common/datavalues/src/types/variant_value.rs

```diff
@@ -0,0 +1,202 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::str::FromStr;
+use std::cmp::Ordering;
+use std::fmt::Display;
+use std::fmt::Formatter;
+use std::ops::Deref;
+
+use common_exception::ErrorCode;
+use common_exception::Result;
+use serde_json::Value;
+
+#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, PartialEq, Eq)]
+pub struct VariantValue(Value);
+
+impl From<Value> for VariantValue {
+    fn from(val: Value) -> Self {
+        VariantValue(val)
+    }
+}
+
+impl From<&Value> for VariantValue {
+    fn from(val: &Value) -> Self {
+        VariantValue(val.clone())
+    }
+}
+
+impl AsRef<Value> for VariantValue {
+    fn as_ref(&self) -> &Value {
+        &self.0
+    }
+}
+
+impl Deref for VariantValue {
+    type Target = Value;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl Default for VariantValue {
+    fn default() -> Self {
+        VariantValue::from(Value::Null)
+    }
+}
+
+impl FromStr for VariantValue {
+    type Err = ErrorCode;
+
+    fn from_str(s: &str) -> Result<Self> {
+        match serde_json::Value::from_str(s) {
+            Ok(v) => Ok(VariantValue::from(v)),
+            Err(_) => Err(ErrorCode::StrParseError(format!(
+                "Unknown VariantValue: {}",
+                s
+            ))),
+        }
+    }
+}
+
+// VariantValue compares as the following rule:
+// Null > Array > Object > String > Number > Boolean
+// The Array compares each element in turn.
+// The Object compares each key-value in turn,
+// first compare the key, and then compare the value if the key is equal.
+// The Greater the key, the Less the Object, the Greater the value, the Greater the Object
+impl Ord for VariantValue {
+    fn cmp(&self, other: &Self) -> Ordering {
+        match (self.as_ref(), other.as_ref()) {
+            (Value::Null, Value::Null) => Ordering::Equal,
+            (Value::Null, _) => Ordering::Greater,
+            (Value::Array(_), Value::Null) => Ordering::Less,
+            (Value::Array(a1), Value::Array(a2)) => {
+                for (v1, v2) in a1.iter().zip(a2) {
+                    if !v1.eq(v2) {
+                        return VariantValue::from(v1).cmp(&VariantValue::from(v2));
+                    }
+                }
+                a1.len().cmp(&a2.len())
+            }
+            (Value::Array(_), _) => Ordering::Greater,
+            (Value::Object(_), Value::Null) => Ordering::Less,
+            (Value::Object(_), Value::Array(_)) => Ordering::Less,
+            (Value::Object(o1), Value::Object(o2)) => {
+                for (k1, k2) in o1.keys().zip(o2.keys()) {
```

Comment:
Is object.keys() sorted ?

Reply:
yes, keys are iterated in alphabetical order.